# Movie Recommendation System

In [531]:
import pandas as pd
import numpy as np
from ast import literal_eval
pd.options.mode.chained_assignment = None

In [566]:
movies = pd.read_csv('movies_metadata.csv', low_memory=False)
credits = pd.read_csv('credits.csv', low_memory=False)
keywords = pd.read_csv('keywords.csv', low_memory=False)

In [567]:
movies.head()

adult                              belongs_to_collection    budget  \
0  FALSE  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  FALSE                                                NaN  65000000   
2  FALSE  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  FALSE                                                NaN  16000000   
4  FALSE  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [568]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  int64  
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [569]:
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [570]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [571]:
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [572]:
keywords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46419 entries, 0 to 46418
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        46419 non-null  int64 
 1   keywords  46419 non-null  object
dtypes: int64(1), object(1)
memory usage: 725.4+ KB


In [573]:
movies = movies.merge(credits, on='id')
movies = movies.merge(keywords, on='id')
movies.head(2)

adult                              belongs_to_collection    budget  \
0  FALSE  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  FALSE                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video  vote_average  \
0                                        NaN  Toy Story  False           7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False           6.9   

   vote_count                                               cast  \
0      5415.0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      2413.0  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  

[2 rows x 27 columns]

In [574]:
movies = movies[['id','title','overview','belongs_to_collection','genres','keywords','cast','crew', 'poster_path']]
movies.head(2)

id      title                                           overview  \
0   862  Toy Story  Led by Woody, Andy's toys live happily in his ...   
1  8844    Jumanji  When siblings Judy and Peter discover an encha...   

                               belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                                            keywords  \
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg

## CONTENT BASED FILTERING
In content based filtering, we will create tags from the available data columns and based on similarity in the tags we will recommend movies to the user.
In this project, content based filtering used based on 
- Overview
- collection
- genres
- keywords
- cast
- director

In [575]:
links_small = pd.read_csv('links_small.csv')
links_small.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9125 entries, 0 to 9124
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  9125 non-null   int64  
 1   imdbId   9125 non-null   int64  
 2   tmdbId   9112 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 214.0 KB


In [576]:

links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype(np.int64)
links_small.info()

<class 'pandas.core.series.Series'>
Int64Index: 9112 entries, 0 to 9124
Series name: tmdbId
Non-Null Count  Dtype
--------------  -----
9112 non-null   int64
dtypes: int64(1)
memory usage: 142.4 KB


In [577]:
small_movies_df = movies[movies['id'].isin(links_small)]
small_movies_df.shape

(9218, 9)

We have 9218 movies avaiable in our small movies metadata dataset which is 5 times smaller than our original dataset of 45000 movies

Since we are using content based filtering in this project, we will have to create tags from the available data columns. To generate tags we will only select relevant columns and discard all of the irrelevant columns from the dataframe.

In [578]:
small_movies_df.head(2)

id      title                                           overview  \
0   862  Toy Story  Led by Woody, Andy's toys live happily in his ...   
1  8844    Jumanji  When siblings Judy and Peter discover an encha...   

                               belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                                            keywords  \
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg

In [579]:
small_movies_df.isnull().sum()

id                          0
title                       0
overview                   12
belongs_to_collection    7530
genres                      0
keywords                    0
cast                        0
crew                        0
poster_path                 3
dtype: int64

In [580]:
small_movies_df['belongs_to_collection'].fillna('{}', inplace=True)

In [581]:
small_movies_df.dropna(inplace=True)

In [582]:
small_movies_df.isnull().sum()
# No missing records

id                       0
title                    0
overview                 0
belongs_to_collection    0
genres                   0
keywords                 0
cast                     0
crew                     0
poster_path              0
dtype: int64

In [583]:
small_movies_df.duplicated().sum()
# 133 records are duplicate and we will drop them

133

In [584]:
small_movies_df.drop_duplicates(inplace=True)

In [585]:
small_movies_df.duplicated().sum()
# Now we have no duplicate records

0

In [586]:
small_movies_df.iloc[0].genres
# genres is a list of dictionaries

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [587]:
def convert_obj(obj):
    list = []
    for i in literal_eval(obj):
        list.append(i['name']) 
    return list

In [588]:
# apply convert_obj function to genres column

small_movies_df['genres'] = small_movies_df['genres'].apply(convert_obj)
small_movies_df['genres'].head()

0     [Animation, Comedy, Family]
1    [Adventure, Fantasy, Family]
2               [Romance, Comedy]
3        [Comedy, Drama, Romance]
4                        [Comedy]
Name: genres, dtype: object

In [589]:
small_movies_df['belongs_to_collection'] = small_movies_df['belongs_to_collection'].apply(lambda x: literal_eval(x)['name'] if x != '{}' else '')
small_movies_df['belongs_to_collection'] = small_movies_df['belongs_to_collection'].apply(lambda x: x.split())
small_movies_df['belongs_to_collection'].head()

0                [Toy, Story, Collection]
1                                      []
2          [Grumpy, Old, Men, Collection]
3                                      []
4    [Father, of, the, Bride, Collection]
Name: belongs_to_collection, dtype: object

In [590]:
small_movies_df['keywords'] = small_movies_df['keywords'].apply(convert_obj)
small_movies_df['keywords'].head(2)

0    [jealousy, toy, boy, friendship, friends, riva...
1    [board game, disappearance, based on children'...
Name: keywords, dtype: object

In [591]:
small_movies_df['overview'] = small_movies_df['overview'].apply(lambda x: x.split())

In [592]:
small_movies_df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                  belongs_to_collection                        genres  \
0              [Toy, Story, Collection]   [Animation, Comedy, Family]   
1                                    []  [Adventure, Fantasy, Family]   
2        [Grumpy, Old, Men, Collection]             [Romance, Comedy]   
3                                    []      [Comedy, Drama, Romance]   
4  [Father, of, the, Bride, Collection]                      [Comedy]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   
2  [fishing, best friend, duringcreditsstinger, o...   
3  [based on novel, interracial relationship, sin...   
4  [baby, midlife crisis, confidence, aging, daug...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg  
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg

In [593]:
def convert_cast(obj):
    list = []
    counter = 0
    for i in literal_eval(obj):
        if counter != 3:
            list.append(i['name'])
            counter += 1
        else:
            break
    return list

In [594]:
small_movies_df['cast'] = small_movies_df['cast'].apply(convert_cast)
small_movies_df.head(2)

id      title                                           overview  \
0   862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  8844    Jumanji  [When, siblings, Judy, and, Peter, discover, a...   

      belongs_to_collection                        genres  \
0  [Toy, Story, Collection]   [Animation, Comedy, Family]   
1                        []  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                             cast  \
0             [Tom Hanks, Tim Allen, Don Rickles]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg

In [595]:
def fetch_director(obj):
    crew = []
    for i in literal_eval(obj):
        if i['job'] == 'Director':
            crew.append(i['name'])
    return crew

In [596]:
small_movies_df['crew'] = small_movies_df['crew'].apply(fetch_director)
small_movies_df.head(2)

id      title                                           overview  \
0   862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  8844    Jumanji  [When, siblings, Judy, and, Peter, discover, a...   

      belongs_to_collection                        genres  \
0  [Toy, Story, Collection]   [Animation, Comedy, Family]   
1                        []  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                             cast             crew  \
0             [Tom Hanks, Tim Allen, Don Rickles]  [John Lasseter]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   [Joe Johnston]   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg

In [597]:
small_movies_df['tags'] = small_movies_df['belongs_to_collection'] + small_movies_df['keywords'] + small_movies_df['genres'] + small_movies_df['overview'] + small_movies_df['cast'] + small_movies_df['crew']

In [620]:
small_movies_df.head(2)

id      title                                           overview  \
0   862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  8844    Jumanji  [When, siblings, Judy, and, Peter, discover, a...   

      belongs_to_collection                        genres  \
0  [Toy, Story, Collection]   [Animation, Comedy, Family]   
1                        []  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                             cast             crew  \
0             [Tom Hanks, Tim Allen, Don Rickles]  [John Lasseter]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   [Joe Johnston]   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                                tags  
0  toy story collection jealousy toy boy friendsh...  
1  board game disappearance based on children's b...

In [599]:
small_movies_df['tags'] = small_movies_df['tags'].apply(lambda x: " ".join(x))
small_movies_df['tags'] = small_movies_df['tags'].apply(lambda x: x.lower())
small_movies_df.head(2)

id      title                                           overview  \
0   862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  8844    Jumanji  [When, siblings, Judy, and, Peter, discover, a...   

      belongs_to_collection                        genres  \
0  [Toy, Story, Collection]   [Animation, Comedy, Family]   
1                        []  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                             cast             crew  \
0             [Tom Hanks, Tim Allen, Don Rickles]  [John Lasseter]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   [Joe Johnston]   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                                tags  
0  toy story collection jealousy toy boy friendsh...  
1  board game disappearance based on children's b...

In [600]:
small_movies_df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                  belongs_to_collection                        genres  \
0              [Toy, Story, Collection]   [Animation, Comedy, Family]   
1                                    []  [Adventure, Fantasy, Family]   
2        [Grumpy, Old, Men, Collection]             [Romance, Comedy]   
3                                    []      [Comedy, Drama, Romance]   
4  [Father, of, the, Bride, Collection]                      [Comedy]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   
2  [fishing, best friend, duringcreditsstinger, o...   
3  [based on novel, interracial relationship, sin...   
4  [baby, midlife crisis, confidence, aging, daug...   

                                                cast               crew  \
0                [Tom Hanks, Tim Allen, Don Rickles]    [John Lasseter]   
1     [Robin Williams, Jonathan Hyde, Kirsten Dunst]     [Joe Johnston]   
2         [Walter Matthau, Jack Lemmon, Ann-Margret]    [Howard Deutch]   
3  [Whitney Houston, Angela Bassett, Loretta Devine]  [Forest Whitaker]   
4         [Steve Martin, Diane Keaton, Martin Short]    [Charles Shyer]   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                                tags  
0  toy story collection jealousy toy boy friendsh...  
1  board game disappearance based on children's b...  
2  grumpy old men collection fishing best friend ...  
3  based on novel interracial relationship single...  
4  father of the bride collection baby midlife cr...

In [601]:
movie_tags = small_movies_df[['id', 'title', 'tags', 'poster_path']]
movie_tags.fillna('', inplace=True)
movie_tags.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                                tags  \
0  toy story collection jealousy toy boy friendsh...   
1  board game disappearance based on children's b...   
2  grumpy old men collection fishing best friend ...   
3  based on novel interracial relationship single...   
4  father of the bride collection baby midlife cr...   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg  
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg

In [602]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(max_features=5000, stop_words='english')

In [603]:
movie_tags.iloc[0].tags

"toy story collection jealousy toy boy friendship friends rivalry boy next door new toy toy comes to life animation comedy family led by woody, andy's toys live happily in his room until andy's birthday brings buzz lightyear onto the scene. afraid of losing his place in andy's heart, woody plots against buzz. but when circumstances separate buzz and woody from their owner, the duo eventually learns to put aside their differences. tom hanks tim allen don rickles john lasseter"

In [604]:
vectors = cv.fit_transform(movie_tags['tags']).toarray()
vectors.shape

(9070, 5000)

In [605]:
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity

In [606]:
cosine_similarity = cosine_similarity(vectors)

In [607]:
cosine_similarity = linear_kernel(vectors, vectors)
cosine_similarity[0]

array([76.,  4.,  5., ...,  0.,  2.,  2.])

In [608]:
movie_tags = movie_tags.reset_index()
movie_tags.head()

index     id                        title  \
0      0    862                    Toy Story   
1      1   8844                      Jumanji   
2      2  15602             Grumpier Old Men   
3      3  31357            Waiting to Exhale   
4      4  11862  Father of the Bride Part II   

                                                tags  \
0  toy story collection jealousy toy boy friendsh...   
1  board game disappearance based on children's b...   
2  grumpy old men collection fishing best friend ...   
3  based on novel interracial relationship single...   
4  father of the bride collection baby midlife cr...   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg  
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg

In [609]:
titles = movie_tags['title']
indices = pd.Series(movie_tags.index, index=movie_tags['title'])

In [610]:
def get_recommendation(title):
    i = indices[title]
    similarity_scores = list(enumerate(cosine_similarity[i]))
    similarity_scores = sorted(similarity_scores, key=lambda x: x[1], reverse=True)
    similarity_scores = similarity_scores[1:6]
    rec_movies = [i[0] for i in similarity_scores]
    return titles.iloc[rec_movies]

In [611]:
get_recommendation('The Godfather')

3708                  3 Ninjas
970     The Godfather: Part II
6136             Batman Begins
962                 GoodFellas
6892           The Dark Knight
Name: title, dtype: object

In [612]:
get_recommendation('The Dark Knight')

7920    The Dark Knight Rises
6136            Batman Begins
1110           Batman Returns
132            Batman Forever
1234           Batman & Robin
Name: title, dtype: object

In [613]:
get_recommendation('Jumanji')

8142           Wreck-It Ralph
6384               Stay Alive
3141       Digimon: The Movie
8152    Indie Game: The Movie
6315            Grandma's Boy
Name: title, dtype: object

In [614]:
get_recommendation('The Lion King')

41                         Restoration
4617                          King Rat
5863    The Lion King 2: Simba's Pride
9062     Kingsglaive: Final Fantasy XV
1865                         King Kong
Name: title, dtype: object

In [615]:
get_recommendation('Fast & Furious')

7766                   Fast Five
2958       Gone in Sixty Seconds
3446    The Fast and the Furious
8289            Fast & Furious 6
4601            2 Fast 2 Furious
Name: title, dtype: object

In [617]:
import pickle

In [618]:
pickle.dump(movie_tags, open('movies.pkl', 'wb'))

In [619]:
pickle.dump(cosine_similarity, open('cosine_similarity.pkl', 'wb'))